core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -605,6 +607,84 @@ public FileSystemMasterView getFileSystemMasterView() {
   }
 
   /**
+   * Checks the consistency of the subtree under the path.
+   *
+   * @param path the subtree root to check
+   * @param options the options to use for the checkConsistency command
+   * @return a list of paths in Alluxio which are not consistent with the under storage
+   * @throws FileDoesNotExistException if the path does not exist
+   * @throws InvalidPathException if the path is invalid
+   * @throws IOException if an error occurs interacting with the under storage
+   */
+  public List<AlluxioURI> checkConsistency(AlluxioURI path, CheckConsistencyOptions options)
+      throws FileDoesNotExistException, InvalidPathException, IOException {
+    List<AlluxioURI> inconsistentUris = new ArrayList<>();
+    Stack<LockedInodePath> pathsToCheck = new Stack<>();
+    try {
+      pathsToCheck.push(mInodeTree.lockInodePath(path, InodeTree.LockMode.READ));
+      while (!pathsToCheck.empty()) {
+        try (LockedInodePath currentPath = pathsToCheck.pop()) {
+          Inode currentInode = currentPath.getInode();
+          if (!checkConsistencyInternal(currentPath)) {
+            inconsistentUris.add(currentPath.getUri());
+          }
+          if (currentInode.isDirectory()) {
+            for (Inode child : ((InodeDirectory) currentInode).getChildren()) {
+              pathsToCheck
+                  .push(mInodeTree.lockFullInodePath(child.getId(), InodeTree.LockMode.READ));
+            }
+          }
+        }
+      }
+    } finally {
+      while (!pathsToCheck.empty()) {
+        pathsToCheck.pop().close();
+      }
+    }
+    return inconsistentUris;
+  }
+
+  /**
+   * Check if a path is consistent between Alluxio and the underlying storage.
+   *
+   * A path without a backing under storage is always consistent.
+   *
+   * A not persisted path is considered consistent if:
+   *   1. It does not shadow an object in the underlying storage.
+   *
+   * A persisted path is considered consistent if:
+   *   1. An equivalent object exists for the its under storage path.
+   *   2. The metadata of the Alluxio and under storage object are equal.
```

Comment:
For now I just put file size, the set of metadata we try to keep in sync will also depend on what we can assume is available from the UFS.